cpu/fe310/include/irq_arch.h

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2017 Ken Rabold
+ *               2020 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup         cpu_fe310
+ * @{
+ *
+ * @file
+ * @brief           Implementation of the kernels irq interface
+ *
+ * @author          Ken Rabold
+ * @author          Alexandre Abadie <alexandre.abadie@inria.fr>
+ */
+
+#ifndef IRQ_ARCH_H
+#define IRQ_ARCH_H
+
+#include <stdint.h>
+#include "irq.h"
+#include "cpu_conf.h"
+#include "cpu.h"
+
+#include "vendor/encoding.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+extern volatile int fe310_in_isr;
+
+/**
+ * @brief Enable all maskable interrupts
+ */
+static inline __attribute__((always_inline)) unsigned int irq_enable(void)
+{
+    /* Enable all interrupts */
+    unsigned retval = read_csr(mstatus);
+    set_csr(mstatus, MSTATUS_MIE);
+    return retval;
+}
+
+/**
+ * @brief Disable all maskable interrupts
+ */
+static inline __attribute__((always_inline)) unsigned int irq_disable(void)
+{
+    unsigned int state = read_csr(mstatus);
+
+    /* Disable all interrupts */
+    clear_csr(mstatus, MSTATUS_MIE);
```

Comment:
```suggestion
    unsigned int state;
    ___asm__ volatile(
        "csrrc %[dest], mstatus, %[mask]"
        : [dest]    "=r"(state)
        : [mask]    "i"(MSTATUS_MIE)
        : "memory"
    );
```

The `csrrc` instruction used by `clear_csr()` not only clears bits, but also returns the old value of the register. But sadly `clear_csr()` just throws this value away. Additionally, `clear_csr()` does not use memory barriers.

Writing the assembly by hand allows to only use one instruction for both disabling IRQs and getting the previous state. Additionally, we can add the memory barrier. I hope that this barrier fixes the regression.

Reply:
Above statement is based on [this doc of `csrrc`](https://msyksphinz-self.github.io/riscv-isadoc/html/rvi.html#csrrc).